src/ray/core_worker/test/core_worker_test.cc

```diff
@@ -560,11 +557,43 @@ void CoreWorkerTest::TestStoreProvider(StoreProviderType type) {
 
   std::thread async_thread(thread_func);
 
-  // wait for the objects to appear.
+  wait_ids.clear();
+  wait_ids.insert(ready_ids.begin(), ready_ids.end());
+  wait_ids.insert(unready_ids.begin(), unready_ids.end());
   wait_results.clear();
+
+  // Check that only the ready ids are returned when timeout ends before thread runs.
   RAY_CHECK_OK(
-      provider.Wait(unready_ids, unready_ids.size(), -1, RandomTaskId(), &wait_results));
-  // wait for the thread to finish.
+      provider.Wait(wait_ids, ready_ids.size() + 1, 100, RandomTaskId(), &wait_results));
+  ASSERT_EQ(ready_ids.size(), wait_results.size());
+  for (const auto &ready_id : ready_ids) {
+    ASSERT_TRUE(wait_results.find(ready_id) != wait_results.end());
+  }
+  for (const auto &unready_id : unready_ids) {
+    ASSERT_TRUE(wait_results.find(unready_id) == wait_results.end());
+  }
+
+  wait_results.clear();
+  // Check that enough objects are returned after the thread inserts at least one object.
+  RAY_CHECK_OK(
+      provider.Wait(wait_ids, ready_ids.size() + 1, 5000, RandomTaskId(), &wait_results));
+  ASSERT_TRUE(wait_results.size() >= ready_ids.size() + 1);
+  for (const auto &ready_id : ready_ids) {
+    ASSERT_TRUE(wait_results.find(ready_id) != wait_results.end());
+  }
+
+  wait_results.clear();
+  // Check that all objects are returned after the thread completes.
```

Comment:
you could move `async_thread.join()` to here to verify the case when the thread completes.

Reply:
Sure, good idea